### ISSUE TITLE
Proposal: Pushing blob chunks could have a max chunk size hint

### ISSUE THREAD

**Issue by gabivlj** [opencontainers/distribution-spec#485]
### Description

Hi! We've encountered a use-case we'd love to discuss over here. We're proposing a new header that could be useful for registries that are behind proxies. Those proxies usually buffer requests so for big layers it degrades performance and memory usage of the server. It would be nice if the registry, either on upload creation (`POST /v2/:name/blobs/uploads/`) or in another way to hint the client the preferred max chunk size.

For example, I've seen places that document the header `OCI-Chunk-Min-Length`, the purpose is hinting a minimum chunk size for clients. It would be amazing to see the counter-part `OCI-Chunk-Max-Length`.

Is it something interesting to see in https://distribution.github.io/distribution/spec/api/? 

I feel like this would be incredibly useful for registries that have limited resources.

**Comment by sudo-bmitch** [comment]
Does this affect Docker's streaming method for a blob push (not defined by OCI yet, but since it's used by docker it's well supported by registries), non-chunked pushes (used by a lot of other tools), or the final chunk size?

**Comment by gabivlj** [comment]
This affects mainly the POST and PATCH method on chunked uploads. It would work the same as `OCI-Chunk-Min-Length`, but by defining a maximum chunk size to send in PATCH to not overwhelm the server.

**Comment by sudo-bmitch** [comment]
Docker's streaming API uses a patch request without a content-length value. It's effectively the same as a post/put, but the digest is not known in advance, so they push the full blob with a single patch.

How do proxies deal with very large blobs uploaded with a post/put?

**Comment by gabivlj** [comment]
> Docker's streaming API uses a patch request without a content-length value. It's effectively the same as a post/put, but the digest is not known in advance, so they push the full blob with a single patch.
> 
> How do proxies deal with very large blobs uploaded with a post/put?

For proxies or platforms like Workers they buffer the request body. So it would be nice to tell the client (like Docker) to limit the request body to a certain size, as there might be a limit in the server on how many bytes it can buffer.

**Comment by sudo-bmitch** [comment]
> For proxies or platforms like Workers they buffer the request body. So it would be nice to tell the client (like Docker) to limit the request body to a certain size, as there might be a limit in the server on how many bytes it can buffer. 

I think that makes my point. The proposed change fixes one of 4 blob upload methods, only 3 of which are defined by OCI, and it's the least used method due to poor support. So I'm unsure how this will solve the underlying issue for proxies.

**Comment by gabivlj** [comment]
> > For proxies or platforms like Workers they buffer the request body. So it would be nice to tell the client (like Docker) to limit the request body to a certain size, as there might be a limit in the server on how many bytes it can buffer. 
> 
> I think that makes my point. The proposed change fixes one of 4 blob upload methods, only 3 of which are defined by OCI, and it's the least used method due to poor support. So I'm unsure how this will solve the underlying issue for proxies.
> 

Please correct me if my understanding here is incorrect, but what I'm proposing is that we do this for all workflows that involve the POST and PATCH and then PUT uploads. I think that involves the chunked upload and the streaming way of Docker as we would also implement in Docker a way for it to respect the limit. 

Monolithic uploads and POST PUT only uploads are the only ones that we are not trying to support here.

**Comment by sudo-bmitch** [comment]
I believe ordered by popularity, there's:

1. POST/PUT
2. Docker's streaming PATCH
3. Chunked POST/PATCH/PUT
4. Single POST

The only method covered by this PR would be 3. OCI hasn't standardized 2 so I don't see how this would apply to that. And if we did define the streaming PATCH, I believe it would have to be excluded from this by its definition.

And 1, being one of the more popular methods, would continue to be an issue for proxies. Given the number of issues I had with registries when I implemented 3, I suspect it's a very small percentage of blob uploads (I'd guess less than 1%). I never tried to implement 4, but I'm guessing a lot of registries using an external blob store would have support issues with that.

Given that close to 100% of blob uploads wouldn't be affected by this change, I'm worried the change to the spec won't have the desired result for your use case.

**Comment by gabivlj** [comment]
> I believe ordered by popularity, there's:
> 
> 
> 
> 1. POST/PUT
> 
> 2. Docker's streaming PATCH
> 
> 3. Chunked POST/PATCH/PUT
> 
> 4. Single POST
> 
> 
> 
> The only method covered by this PR would be 3. OCI hasn't standardized 2 so I don't see how this would apply to that. And if we did define the streaming PATCH, I believe it would have to be excluded from this by its definition.
> 
> 
> 
> And 1, being one of the more popular methods, would continue to be an issue for proxies. Given the number of issues I had with registries when I implemented 3, I suspect it's a very small percentage of blob uploads (I'd guess less than 1%). I never tried to implement 4, but I'm guessing a lot of registries using an external blob store would have support issues with that.
> 
> 
> 
> Given that close to 100% of blob uploads wouldn't be affected by this change, I'm worried the change to the spec won't have the desired result for your use case.

Interesting. I think this would benefit us as we would have a justification for a Docker PR where we support this functionality for the 2nd use case (see PR mentioned on top). We are also fine supporting only 3rd case outside of Docker. We are not targeting 1st and 4th at the moment.

**Comment by sudo-bmitch** [comment]
For reference, the streaming patch is defined in the [distribution project](https://distribution.github.io/distribution/spec/api/#stream-upload). In practice, this was implemented as a POST / single PATCH (no length or range header) / PUT. I believe this is done for packing the layer content on the fly, where the output of the tar + compress of the files is not known in advance and they didn't want to write to a temp file or memory.

Supporting more than one PATCH in a stream request is interesting, and looks like it's technically allowed in the distribution project spec, even if it was not previously implemented that way in practice. Since OCI hasn't defined that API yet, I'd be curious to hear from registry operators if they have any concerns with supporting multiple PATCH requests on that API (trying to think ahead for if/when we do add that API).

**Comment by gabivlj** [comment]
To clarify nomenclature, stream request you mean the way docker pushes a layer to PATCH?

From what I've seen in the code, I think it's doable to have multiple PATCH calls from Docker if we add a limit reader like here:
https://github.com/distribution/distribution/pull/4144

What I want to get here is just have a way for clients/registries to agree on a max request body size for more constrained environments, seems like a good place to put this is what I'm proposing here:
https://github.com/opencontainers/distribution-spec/pull/486

**Comment by sudo-bmitch** [comment]
Yes, I believe docker was using the API that I linked above. Changing the behavior of the docker client or the distribution registry should be easy. Making sure that such a change doesn't impact any other registry implementations (ecr, gar, acr, quay, harbor, zot, etc) is where I'm looking for feedback from registry operators. It also means OCI, if/when we added that API, would need to support a looser implementation where multiple PATCH requests without a range or length would be allowed.

**Comment by gabivlj** [comment]
I'm looking to document this and making it an optional field for registries operators to implement, I think it wouldn't be a breaking change for existing registries AFAIK?

**Comment by gabivlj** [comment]
I also want to add that this seems similar to https://github.com/opencontainers/distribution-spec/pull/391. Who can chime in this discussion that could review the change and bring more to the discussion?

Thanks for the comments on this @sudo-bmitch btw, I appreciate the input on this!

**Comment by sudo-bmitch** [comment]
The difference between the two PRs is the problem being solved with the feature, which APIs it applies to, and how that change will affect existing implementations.

#391 was focused on just the chunked upload, which is relatively unused, and only needed to fix a problem with how that API interacted with backends. Since it was so unused, there was little impact to clients, and the only reason it was noticed at all was because the conformance tests used unusually small chunk sizes.

This PR is trying to standardize a change on how every Docker push is run since it's including the streaming API, which isn't even part of the OCI spec. And it's trying to fix a problem with proxies, which will only be solved for those using specific APIs.

As a result, it's both a much larger change, that may affect existing registries, and one which I expect the partial implementation will result in users complaining of broken tooling depending on which push method is used by their specific tool (I've looked at more than a few requests from users complaining about broken tools when the issue is tracked down to something on their network). So I believe waiting for feedback from the community would be valuable in this scenario.

**Comment by gabivlj** [comment]
I agree that a change like this should be methodical. However, this code path is only activated by the few registries that implement this OCI-Chunked-Max-Upload header. 
Docker will keep doing the same path of upload in a single patch the layer. If that is still not OK I understand, but it's a code path that is only activated by registries that need it.
Maybe we should advise that it's an experimental feature or something the registry should avoid if it cans?

**Comment by sudo-bmitch** [comment]
Could it be activated by a user controlled http proxy on all outbound connections from their network, going to a general cloud registry?

**Comment by gabivlj** [comment]
The idea here is that this is mainly designed for serverless platforms like Workers or registries that are hosted by the user behind a proxy that buffers requests. 
These platforms have the limitation that the request body cannot be bigger than X. This is configured by the user that owns the registry, so the path is activated only by the user that maintains the registry. Shouldn't be the default by any case for normal registries that dont have these restrictions.

**Comment by sudo-bmitch** [comment]
How does this prevent networks with a forward proxy from adding this header to responses from third party registries? Wouldn't those networks also want the ability to limit the size of a blob chunks so they can inspect the content before it leaves their network? That's a fairly common network model for large organizations with security or regulation requirements.

**Comment by gabivlj** [comment]
Sure! That's also a legitimate use-case. I don't see a problem with that. I feel like that's also an opt-in behaviour from the users of the registries.

**Comment by sudo-bmitch** [comment]
The user with an arbitrary client tool, the network operator of the proxy, and the registry owner, are typically 3 different groups in environments I've seen. In those scenarios, the client tooling didn't opt into it, the registry operator didn't either, and the user will often report issues to both of them rather than work with their own network team.

**Comment by gabivlj** [comment]
If the client didn't opt-in (chooses to ignore the header) then it will work the same as always. The behaviour will be the same as before because the proxy will still buffer the request. If it chooses to opt-in to the behaviour (implements the header) it will chunk the request.

**Comment by sudo-bmitch** [comment]
I'm going to take a step back from this issue because I don't feel my concerns are being addressed and I don't think the people that should be addressing them are going to read the full discussion.

**Comment by gabivlj** [comment]
Hi, sorry to circle back on this. Is there any folks that are able to add more context to discussion? Thank you!

**Comment by agahkarakuzu** [comment]
Thanks for the prompt response @gabivlj! 

> from my previous tests with it I've seen it try to push a layer in a single chunk no matter what's the size.

I guess this addition would enforce a chunked upload? Is there an approximate timeline to merge it? I guess my only option to keep the proxy on, then turn it off to save the day once in a while.

**Comment by harris2012** [comment]
considering someone is pushing a very large image like 1000M. 

Without the `OCI-Chunk-Max-Length`.
any little network failure will let him patch whole image from 0%.

With the `OCI-Chunk-Max-Length`.
He can patch the image every 10M. When networkle failure at 40%, he will continue to retry from 40%.

**Comment by sudo-bmitch** [comment]
> considering someone is pushing a very large image like 1000M. 
> 
> Without the `OCI-Chunk-Max-Length`.
> any little network failure will let him patch whole image from 0%.
> 
> With the `OCI-Chunk-Max-Length`.
> He can patch the image every 10M. When networkle failure at 40%, he will continue to retry from 40%.
> 

That's supported today with chunked uploads. But few use chunked uploads (they are a lot slower with the added round trips, less supported by registries, and more complicated # to implement). So this issue wouldn't help anyone experiencing the issue you describe.

**Comment by smashah** [comment]
With more and more people using tools like cloudflare tunnels to manage their routing, this is more needed than ever since it appears to be the blocker for https://github.com/distribution/distribution/pull/4144 and therefore a blocker for https://github.com/docker/cli/issues/5589

It's probably just an edge case today but more people will be coming across this after hours of time wasted in the future.

What is blocking this?

**Comment by mikebrow** [comment]
@dmcgowan FYI

**Comment by laurentpellegrino** [comment]
Any progress?

**Comment by omartrigui** [comment]
Any updates?
